### PR TITLE
ESXi Powershell and API development

### PR DESCRIPTION
**Pull request by warku123**
In this PR, I developed test cases by ESXi API first. But found that some of the function is not sufficent.
So I tried use linux ESXi PowerCLI then.

The file using ESXi API is not available, but the other file can run well.

This PR may used in the future, but not now.

### CODE REVIEW

File: hypervisor/virt/esx/esxapi.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+class Esxapi:
+    def __init__(
+        self,
+        server,
+        user,
+        passwd,):
+        """
+        Collect vcenter information, provide Host add/delete and
+        Guest add/delete/start/stop/suspend/resume functions
+        via ESXi official API
```
Comment (hkx303) [inline hypervisor/virt/esx/esxapi.py:16]: can add `following document:  [api doc link]` here

File: hypervisor/virt/esx/esxapi.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            # "esx_cpu": str(output["VMHost"]["NumCpu"]),
+            # "esx_cluster": output["VMHost"]["Parent"],
+        }
+        if uuid_info:
+            guest_msgs["guest_uuid"] = self.guest_uuid(guest_name)
+            guest_msgs["esx_uuid"] = self.host_uuid(host_name)
+            guest_msgs["esx_hwuuid"] = self.host_hwuuid(host_name)
+        return guest_msgs
+    
+    def guest_start(self, guest_name):
```
Comment (hkx303) [inline hypervisor/virt/esx/esxapi.py:138]: doc string please :)

File: hypervisor/virt/esx/esxapi.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        )
+        
+        if response.status_code==204:
+            logger.info("Succeeded to start vcenter guest")
+            return True
+        else:
+            logger.error("Failed to start vcenter guest")
+            return False
+    
+    def guest_stop(self, guest_name):
```
Comment (hkx303) [inline hypervisor/virt/esx/esxapi.py:162]: doc string

File: hypervisor/virt/esx/linux_powercli.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    
+    def _format(self, ret=0, stdout=None):
+        """
+        Convert the json string to python list data
+        :param ret: return code
+        :param stdout: output for the execute command
+        :return: the list after json.loads
+        """
+        # regex = r"[[][.\s\S]*[]]"
+        # res = re.findall(regex, stdout, re.DOTALL)
```
Comment (hkx303) [inline hypervisor/virt/esx/linux_powercli.py]: you could delete these 2 lines directly

File: hypervisor/virt/esx/linux_powercli.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+        if ret == 0 and res is not None:
+            return json.loads(res)
+
+    def info(self):
+        """
+        Get the VMhost info
+        :return: the VMHosts info
+        """
+        cmd = f"{self.cert} ConvertTo-Json @(Get-VMHost | Select * -ExcludeProperty ExtensionData)"
+        cmd += f'\''
```
Comment (hkx303) [inline hypervisor/virt/esx/linux_powercli.py]: I am sorry, why we need this line?

File: hypervisor/virt/esx/linux_powercli.py [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+        :param host_ssh_pwd: the ssh password for the host
+        :param guest_name: the name for the new virtual machine
+        :param image_path: the path for the guest image which from the remote web
+        :return: create successfully, return True, else, return False
+        """
+        if self.guest_images(host, host_ssh_user, host_ssh_pwd, guest_name, image_path):
+            data_store = self.host_search(host)["host_data_store"]
+            vmxFile = f"'[{data_store}] {guest_name}/{guest_name}.vmx'"
+            cmd = f"{self.cert} New-VM -VMFilePath {vmxFile} -VMHost {host}"
+            cmd += f'\''
```
Comment (hkx303) [inline hypervisor/virt/esx/linux_powercli.py]: the same

**Comment by hkx303** [comment]
Glad to see we have Linux powerCLI support, I have added some small comments, and could you also add the test result for this PR, especially for the linux_powercli python file, like add the output of the function `guest_search()`, thanks

**Comment by warku123** [comment]
``` python
# test result
if __name__ == "__main__":
    cli = PowerCLI(
        server = '10.73.213.123',
        admin_user = 'administrator@vsphere.local',
        admin_passwd = 'Welcome1!',
        client_server='10.73.5.223',
        client_user = 'root',
        client_passwd = 'red2015'
    )
    print(cli.guest_search("8.7_Server_x86_64-85"))
```
Result here:
```
{'guest_name': '8.7_Server_x86_64-85', 'guest_ip': '10.73.212.245', 'guest_state': 1, 'guest_cpu': '1', 'esx_ip': '10.73.213.144', 'esx_hostname': '10.73.213.144', 'esx_version': '8.0.0', 'esx_state': 1, 'esx_cpu': '4', 'esx_cluster': 'Cluster-Virtwho-85'}
```

File: hypervisor/virt/esx/linux_powercli.py [hunk 7]
```diff
@@ -45,20 +45,20 @@ def _format(self, ret=0, stdout=None):
...
-        # regex = r"[[][.\s\S]*[]]"
-        # res = re.findall(regex, stdout, re.DOTALL)
         stdout = re.sub('\[+[\d+$]', '', stdout)
         res = re.findall(r"[[][\W\w]+[]]", stdout)[0]
         if ret == 0 and res is not None:
             return json.loads(res)
 
+    def cmd_format(self, cmd):
+        return self.cert+cmd+'\''
+
```
Comment (hkx303) [inline hypervisor/virt/esx/linux_powercli.py:52]: Seems we don't know need this function

**Review by hkx303 — APPROVED** [review]

**Review by hkx303 — APPROVED** [review]

**Comment by hkx303** [comment]
@warku123 pls remove the ip username and address in https://github.com/VirtwhoQE/hypervisor-builder/pull/33#issuecomment-1643456819 as these are private, thanks :)

File: hypervisor/virt/esx/esxapi.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        # https://developer.vmware.com/apis/vsphere-automation/latest/vcenter/vm/guest.power/
```
Comment (hkx303) [inline hypervisor/virt/esx/esxapi.py:163]: missed doc string, like:
```
        """
        Power off virtual machines.
        :param guest_name: the virtual machines you want to power off.
        :return: stop successfully, return True, else, return False.
        """
```

Comment (hkx303) [reply hypervisor/virt/esx/esxapi.py:163]: but it is OK, as we cannot use the API for now

**Review by hkx303 — APPROVED** [review]

**Pull request merged**